### PR TITLE
Implement risc0 adapter

### PR DESCRIPTION
**Pull request by preston-evans98**

### CODE REVIEW

File: risc0-adapter/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "risc0-adapter"
```
Comment (bkolad) [inline risc0-adapter/Cargo.toml:2]: Doe is make sense to create `adapters/risc0` , this way there will be a natural place to add other adapters?

File: risc0-adapter/src/host.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+impl<'prover> Zkvm for Risc0Host<'prover> {
+    type CodeCommitment = Risc0MethodId;
+
+    type Error = anyhow::Error;
+
+    fn verify<'a>(
+        serialized_proof: &'a [u8],
+        code_commitment: &Self::CodeCommitment,
+    ) -> Result<&'a [u8], Self::Error> {
+        // let journal:  = bincode::deserialize(serialized_proof)?;
```
Comment (citizen-stig) [inline risc0-adapter/src/host.rs:46]: Let's remove commented code.

File: risc0-adapter/src/host.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            ),
+            &code_commitment.0,
+            &receipt.seal,
+            receipt.journal,
+        )?;
+        Ok(receipt.journal)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
```
Comment (citizen-stig) [inline risc0-adapter/src/host.rs:60]: According to #285 we prefere fully qualified serde derive, so you don't need to feature gate import.

File: risc0-adapter/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+pub mod guest;
+#[cfg(feature = "native")]
+pub mod host;
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Risc0MethodId([u32; 8]);
+
+impl Matches<Self> for Risc0MethodId {
+    fn matches(&self, other: &Self) -> bool {
+        self.0 == other.0
```
Comment (citizen-stig) [inline risc0-adapter/src/lib.rs:14]: Why not implement automatic implementation for `PartialEq` types?

```rust
impl<T: PartialEq> Matches<T> for T {
    fn matches(&self, other: &T) -> bool {
        self == other
    }
}
```

File: risc0-adapter/Cargo.toml (hunk 1 above)
Comment (preston-evans98) [reply risc0-adapter/Cargo.toml:2]: Good idea. Moved!

File: risc0-adapter/src/host.rs (hunk 2 above)
Comment (preston-evans98) [reply risc0-adapter/src/host.rs:46]: Oops, good catch. Removed!

File: risc0-adapter/src/host.rs (hunk 3 above)
Comment (preston-evans98) [reply risc0-adapter/src/host.rs:60]: This package can't feature gate the import, so we don't need to qualify the paths to make it easier. I've qualified it just to make it uniform with the other packages though

File: risc0-adapter/src/lib.rs (hunk 4 above)
Comment (preston-evans98) [reply risc0-adapter/src/lib.rs:14]: `Matches` is a looser concept than `PartialEq` - so a type where `a != b`  might still return `true` for `a.matches(b)`.

**Review by bkolad — APPROVED** [review]

**Pull request merged**